### PR TITLE
Migrate codegen to async

### PR DESCRIPTION
**Pull request by marabooy**
Updates codegen to be async and builds on top of #413 
- [x] Migrates the code generator to transform async.
- [x] Uses the Threading model within visual studio.
- [X] eliminates loading a file to see if it is different from the generated file.

### CODE REVIEW

File: test/ODataConnectedService.Tests/CodeGeneration/CodeGenDescriptorTest.cs [hunk 1]
```diff
@@ -187,32 +188,27 @@ public void TestAddGenerateClientCode_GeneratesMultipleFiles()
...
-
             //The file manager expects the files to have been saved in the Temp directory
             // when ODataT4CodeGenerator.TransformText() was called. Since we're using a dummy code generator
             // we need to manually ensure those files exist
-            var file1TempPath = Path.Combine(Path.GetTempPath(), "File1.cs");
-            File.WriteAllText(file1TempPath, "Contents1");
-            var file2TempPath = Path.Combine(Path.GetTempPath(), "File2.cs");
-            File.WriteAllText(file2TempPath, "Contents2");
-
+            codeGen.MultipleFilesManager.GenerateFilesAsync(true).Wait();
```
Comment (habbes) [inline test/ODataConnectedService.Tests/CodeGeneration/CodeGenDescriptorTest.cs:207]: Why not making this test async and use `async/await`? Or is that done in another PR?

**Review by habbes — COMMENTED** [review]
Looks mostly good to me. I noticed some `Wait()` calls in the tests, is that intentional?

File: src/Microsoft.OData.CodeGen/CodeGeneration/V4CodeGenDescriptor.cs [hunk 2]
```diff
@@ -201,7 +201,7 @@ private async Task AddGeneratedCodeAsync(string metadata, string outputDirectory
 
             using (StreamWriter writer = File.CreateText(tempFile))
             {
-                await writer.WriteAsync(t4CodeGenerator.TransformText());
+                await writer.WriteAsync(await t4CodeGenerator.TransformTextAsync());
```
Comment (WanjohiSammy) [inline src/Microsoft.OData.CodeGen/CodeGeneration/V4CodeGenDescriptor.cs:204]: By default, this will wait for the same thread to continue. Is that the expected behavior here

File: src/Microsoft.OData.CodeGen/CodeGeneration/V4CodeGenDescriptor.cs [hunk 3]
```diff
@@ -214,7 +214,10 @@ private async Task AddGeneratedCodeAsync(string metadata, string outputDirectory
 
             var outputFile = Path.Combine(referenceFolder, $"{this.GeneratedFileNamePrefix(serviceConfiguration.GeneratedFileNamePrefix)}{(languageOption == LanguageOption.GenerateCSharpCode ? ".cs" : ".vb")}");
```
Comment (WanjohiSammy) [inline src/Microsoft.OData.CodeGen/CodeGeneration/V4CodeGenDescriptor.cs:215]: There has been recommendation from team member to use explicit types instead of `var`. I am not sure if that should also be applied to ODataConnectedService

File: src/Microsoft.OData.CodeGen/CodeGeneration/V4CodeGenDescriptor.cs (hunk 2 above)
Comment (marabooy) [reply src/Microsoft.OData.CodeGen/CodeGeneration/V4CodeGenDescriptor.cs:204]: Yes this is the expected behaviour and the message handler provides users with a progress dialog when this method is executing.

File: src/Microsoft.OData.CodeGen/CodeGeneration/V4CodeGenDescriptor.cs (hunk 3 above)
Comment (marabooy) [reply src/Microsoft.OData.CodeGen/CodeGeneration/V4CodeGenDescriptor.cs:215]: This project unfortunately uses var in a lot of places and the reason in ODL var is discouraged is so we can catch any breaking changes. Since this is not shipped as a library there is no such concern.

E.g. In ODL 
```
IEnumerable<string> elements() => new string[]{....}

string[] elements() => new string[] {....} // This is flagged in ODL. However for locals we could use var in ODL without many issues but we maintain concrete types so that changes like these are flagged in all places where the function is called.
```

File: test/ODataConnectedService.Tests/CodeGeneration/CodeGenDescriptorTest.cs (hunk 1 above)
Comment (marabooy) [reply test/ODataConnectedService.Tests/CodeGeneration/CodeGenDescriptorTest.cs:207]: This is meant to be done in the STA pr. which this merges to 
image

Comment (marabooy) [reply test/ODataConnectedService.Tests/CodeGeneration/CodeGenDescriptorTest.cs:207]: Let me resolve this in that pr as the changes will be quite long as this pr has a bunch of blocking wait calls.

**Review by WanjohiSammy — APPROVED** [review]

**Review by habbes — APPROVED** [review]
